crates/dht/src/dht/tests.rs

```diff
@@ -0,0 +1,588 @@
+    assert!(dht1.is_in_sync_with(&dht2).await.unwrap());
+
+    // Now put mismatched data in the common sector
+    dht1.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(
+            (2 * SECTOR_SIZE).to_le_bytes().to_vec(),
+        )),
+        UNIX_TIMESTAMP,
+        vec![],
+    )])
+    .await
+    .unwrap();
+    dht2.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(
+            (2 * SECTOR_SIZE + 1).to_le_bytes().to_vec(),
+        )),
+        UNIX_TIMESTAMP,
+        vec![],
+    )])
+    .await
+    .unwrap();
+
+    // Try to sync the DHTs
+    let outcome = dht1.sync_with(&mut dht2).await.unwrap();
+    assert!(matches!(outcome, SyncWithOutcome::SyncedDisc));
+
+    // Sync the ops
+    dht1.apply_op_sync(&mut dht2).await.unwrap();
+
+    // Now the DHTs should be in sync
+    assert!(dht1.is_in_sync_with(&dht2).await.unwrap());
+}
+
+#[tokio::test]
+async fn ring_sync_respects_arc() {
+    let current_time = Timestamp::now();
+    let mut dht1 =
+        DhtSyncHarness::new(current_time, DhtArc::Arc(0, 3 * SECTOR_SIZE - 1))
+            .await;
+    let mut dht2 = DhtSyncHarness::new(
+        current_time,
+        DhtArc::Arc(2 * SECTOR_SIZE, 4 * SECTOR_SIZE - 1),
+    )
+    .await;
+
+    // Put mismatched recent data in both DHTs, but in sectors that don't overlap
+    dht1.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(SECTOR_SIZE.to_le_bytes().to_vec())),
+        dht1.dht.partition.full_slice_end_timestamp(),
+        vec![],
+    )])
+    .await
+    .unwrap();
+    dht2.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(
+            (3 * SECTOR_SIZE).to_le_bytes().to_vec(),
+        )),
+        dht1.dht.partition.full_slice_end_timestamp(),
+        vec![],
+    )])
+    .await
+    .unwrap();
+
+    // At this point, the DHTs should be in sync, because the mismatch is not in common sectors
+    assert!(dht1.is_in_sync_with(&dht2).await.unwrap());
+
+    // Now put mismatched data in the common sector
+    dht1.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(
+            (2 * SECTOR_SIZE).to_le_bytes().to_vec(),
+        )),
+        dht1.dht.partition.full_slice_end_timestamp(),
+        vec![],
+    )])
+    .await
+    .unwrap();
+    dht2.inject_ops(vec![Kitsune2MemoryOp::new(
+        OpId::from(bytes::Bytes::from(
+            (2 * SECTOR_SIZE + 1).to_le_bytes().to_vec(),
+        )),
+        dht1.dht.partition.full_slice_end_timestamp(),
+        vec![],
+    )])
+    .await
+    .unwrap();
+
+    // Try to sync the DHTs
+    let outcome = dht1.sync_with(&mut dht2).await.unwrap();
+    assert!(matches!(outcome, SyncWithOutcome::SyncedRings));
+
+    // Sync the ops
+    dht1.apply_op_sync(&mut dht2).await.unwrap();
+
+    // Now the DHTs should be in sync
+    assert!(dht1.is_in_sync_with(&dht2).await.unwrap());
+}
```

Comment:
I think something like an initial gossip session that syncs down to the level of hashes would be good to add. Also to get a feel for how that'll work. What do you think?
```suggestion
}

#[tokio::test]
async fn disc_sync() {
    let current_time = Timestamp::now();
    let store_1 = Arc::new(Kitsune2MemoryOpStore::default());
    let mut dht1 = Dht::try_from_store(current_time, store_1.clone())
        .await
        .unwrap();
    let store_2 = Arc::new(Kitsune2MemoryOpStore::default());
    let dht2 = Dht::try_from_store(current_time, store_2.clone())
        .await
        .unwrap();

    let arc_set_1 = ArcSet::new(SECTOR_SIZE, vec![DhtArc::FULL]).unwrap();
    let arc_set_2 = ArcSet::new(SECTOR_SIZE, vec![DhtArc::FULL]).unwrap();

    let op_id_1 = OpId::from(bytes::Bytes::from_static(&[0]));
    let op_1 = Kitsune2MemoryOp::new(op_id_1.clone(), UNIX_TIMESTAMP, vec![]);
    store_1
        .process_incoming_ops(vec![op_1.clone().try_into().unwrap()])
        .await
        .unwrap();
    dht1.inform_ops_stored(store_1.clone(), vec![op_1.into()])
        .await
        .unwrap();

    let dht_minimal_1 = dht1
        .snapshot_minimal(&arc_set_1, store_1.clone())
        .await
        .unwrap();

    let next_action_1 = dht2
        .handle_snapshot(&dht_minimal_1, None, &arc_set_2, store_2.clone())
        .await
        .unwrap();

    let snapshot_disc_sectors_2 = match next_action_1 {
        DhtSnapshotNextAction::NewSnapshot(snapshot) => snapshot,
        _ => panic!("ohono"),
    };

    let next_action_2 = dht1
        .handle_snapshot(
            &snapshot_disc_sectors_2,
            Some(dht_minimal_1),
            &arc_set_1,
            store_1.clone(),
        )
        .await
        .unwrap();

    let snapshot_disc_sector_details_1 = match next_action_2 {
        DhtSnapshotNextAction::NewSnapshot(snapshot) => snapshot,
        _ => panic!("ohono"),
    };

    let next_action_1 = dht2
        .handle_snapshot(
            &snapshot_disc_sector_details_1,
            Some(snapshot_disc_sectors_2),
            &arc_set_2,
            store_2.clone(),
        )
        .await
        .unwrap();

    let snapshot_and_hash_list_2 = match next_action_1 {
        DhtSnapshotNextAction::NewSnapshotAndHashList(snapshot, hashes) => {
            (snapshot, hashes)
        }
        _ => panic!("ohono"),
    };

    let next_action_2 = dht1
        .handle_snapshot(
            &snapshot_and_hash_list_2.0,
            Some(snapshot_disc_sector_details_1),
            &arc_set_1,
            store_1.clone(),
        )
        .await
        .unwrap();

    let hashes = match next_action_2 {
        DhtSnapshotNextAction::HashList(hashes) => hashes,
        _ => panic!("ohono"),
    };

    assert_eq!(hashes, vec![op_id_1]);
}
```